src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java

```diff
@@ -0,0 +1,112 @@
+    private final String memTypeAttributeKey = "mem_type";
+    private boolean metricsInitialised;
+
+    public RTFHeapMetricsCollector() {
+        super(
+                SAMPLING_TIME_INTERVAL,
+                "RTFHeapMetricsCollector",
+                StatMetrics.HEAP_METRICS_COLLECTOR_EXECUTION_TIME,
+                StatExceptionCode.HEAP_METRICS_COLLECTOR_ERROR);
+        this.metricsInitialised = false;
+    }
+
+    @Override
+    public void collectMetrics(long startTime) {
+        metricsRegistry = OpenSearchResources.INSTANCE.getMetricsRegistry();
+        if (metricsRegistry == null) {
+            LOG.error("could not get the instance of MetricsRegistry class");
+            return;
+        }
+
+        initialiseMetricsIfNeeded();
+        GCMetrics.runGCMetrics();
+        LOG.debug("Executing collect metrics for RTFHeapMetricsCollector");
+        recordMetrics();
+    }
+
+    private void initialiseMetricsIfNeeded() {
+        if (metricsInitialised == false) {
+            gcCollectionEventMetrics =
+                    metricsRegistry.createHistogram(
+                            AllMetrics.HeapValue.Constants.COLLECTION_COUNT_VALUE,
+                            "GC Collection Event PA Metrics",
+                            "");
+
+            gcCollectionTimeMetrics =
+                    metricsRegistry.createHistogram(
+                            AllMetrics.HeapValue.Constants.COLLECTION_TIME_VALUE,
+                            "GC Collection Time PA Metrics",
+                            "");
+
+            heapUsedMetrics =
+                    metricsRegistry.createHistogram(
+                            AllMetrics.HeapValue.Constants.USED_VALUE,
+                            "GC Heap Used PA Metrics",
+                            "");
+            metricsInitialised = true;
+        }
+    }
+
+    private void recordMetrics() {
+        Tags totYoungGCTag =
+                Tags.create()
+                        .addTag(memTypeAttributeKey, AllMetrics.GCType.TOT_YOUNG_GC.toString());
+
+        Tags totFullGCTag =
+                Tags.create().addTag(memTypeAttributeKey, AllMetrics.GCType.TOT_FULL_GC.toString());
+
+        gcCollectionEventMetrics.record(GCMetrics.getTotYoungGCCollectionCount(), totYoungGCTag);
+
+        gcCollectionEventMetrics.record(GCMetrics.getTotFullGCCollectionCount(), totFullGCTag);
+
+        gcCollectionTimeMetrics.record(GCMetrics.getTotYoungGCCollectionTime(), totYoungGCTag);
+
+        gcCollectionTimeMetrics.record(GCMetrics.getTotFullGCCollectionTime(), totFullGCTag);
+
+        for (Map.Entry<String, Supplier<MemoryUsage>> entry :
+                HeapMetrics.getMemoryUsageSuppliers().entrySet()) {
+            MemoryUsage memoryUsage = entry.getValue().get();
+            heapUsedMetrics.record(
+                    memoryUsage.getUsed(),
+                    Tags.create().addTag(memTypeAttributeKey, entry.getKey()));
+            metricsRegistry.createGauge(
```

Comment:
Gauge should also be created once.

Reply:
I tried creating the gauge object once and storing it but it does not provide any method that could allow us to re-use an existing object for measurements.

Reply:
Gauge accepts one Supplier which will be executed once in a defined period which is default 1 min in OpenSearch core.

Reply:
But tags can only be added while creating gauge. That would be a problem in your approach.

Reply:
Discussed offline to keep it as it is until further development